### PR TITLE
CreateVectorizedFunction using only template parameters

### PR DESCRIPTION
**Pull request by tiagokepe**
Hello all,

This PR implements the CreateVectorizedFunction using only template parameters, it's a complement to the UDF C++ API.

It's very similar to the first CreateFunction implemented by the PR #712, the main difference is in the second argument that instead of receiving a generic function pointer (i.e., TR (*udf_func)(Args…)), it receives a vectorized function pointer of the type _scalar_function_t_:

`typedef std::function<void(DataChunk &, ExpressionState &, Vector &)> scalar_function_t;`

**1.** `template<typename TR, typename... Args> void CreateVectorizedFunction(string name, scalar_function_t udf_func)`

- template parameters:
    - **TR** is the return type of the UDF function;
    - **Args** are the arguments up to 3 for the UDF function.
- **name**: is the name to register the UDF function;
- **udf_func**: is a vectorized UDF function.

This function automatically discovers from the template typenames the corresponding SQLTypes:

- bool → SQLType::BOOLEAN;
- int8_t →SQLType::TINYINT;
- int16_t →SQLType::SMALLINT
- int32_t →SQLType::INTEGER
- int64_t  →SQLType::BIGINT
- float →SQLType::FLOAT
- double →SQLType::DOUBLE
- string_t →SQLType::VARCHAR

An example of use would be:

```
static void udf_unary_function(DataChunk &input, ExpressionState &state, Vector &result) {
	assert(input.column_count() == 1);
	assert(input.data[0].type == TypeId::INTEGER);

	result.vector_type = VectorType::FLAT_VECTOR;
	auto result_data = FlatVector::GetData<int>(result);
	auto ldata = FlatVector::GetData<int>(input.data[0]);

	FlatVector::SetNullmask(result, FlatVector::Nullmask(input.data[0]));
	for (idx_t i = 0; i < input.size(); i++) {
		result_data[i] = ldata[i];
	}
}

connection.CreateVectorizedFunction<int, int>("udf_unary_int_function", &udf_unary_function);
```

P.S. It's missing a CreateVectorizedFunction to disambiguate some SQL types (some primitive types, e.g., _int32_t_, are mapped to the same SQLType: INTEGER, TIME and DATE).  Such function will be implemented soon in the same style of the second CreateFunction present in the PR #712, with some extra SQLType arguments.

### CODE REVIEW

File: test/sql/udf_function/test_vectorized_udf_function.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	unique_ptr<QueryResult> result;
+	DuckDB db(nullptr);
+	Connection con(db);
+	con.EnableQueryVerification();
+
+	string func_name, table_name, col_type;
+	//The types supported by the templated CreateVectorizedFunction
+	const vector<SQLType> sql_templated_types = {SQLType::BOOLEAN, SQLType::TINYINT, SQLType::SMALLINT,
+												 SQLType::INTEGER, SQLType::BIGINT, SQLType::FLOAT,
+												 SQLType::DOUBLE}; //, SQLType::VARCHAR
```
Comment (Mytherin) [inline test/sql/udf_function/test_vectorized_udf_function.cpp:18]: Why is the VARCHAR commented out here?

File: test/sql/udf_function/test_vectorized_udf_function.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
+#include "catch.hpp"
+#include "test_helpers.hpp"
+#include "udf_functions_to_test.hpp"
+
+using namespace duckdb;
+using namespace std;
+
+TEST_CASE("Vectorized UDF functions", "[udf_function]") {
```
Comment (Mytherin) [inline test/sql/udf_function/test_vectorized_udf_function.cpp:8]: Could you add a test with 4 or more input columns? Does not need to be for every type, one function will suffice.

File: src/include/duckdb/main/connection.hpp [hunk 3]
```diff
@@ -115,6 +115,11 @@ class Connection {
 		UDFWrapper::RegisterFunction(name, args, ret_type, function, *context);
 	}
 
+	template<typename TR, typename... Args>
+	void CreateVectorizedFunction(string name, scalar_function_t udf_func) {
```
Comment (Mytherin) [inline src/include/duckdb/main/connection.hpp:119]: Could you add support here for vararg functions as well and add a test case for them?

**Review by Mytherin — COMMENTED** [review]
Hey Tiago, thanks for the PR again! Some minor comments and requests :)

**Comment by tiagokepe** [comment]
I think this PR is ready, just the R package tests didn't pass due to the error: "Can't find DAV_PASSWORD in env".

**Comment by Mytherin** [comment]
Indeed, this is ready to merge! Thanks for the PR again :)

**Pull request merged**